R/mock-session.R

```diff
@@ -0,0 +1,283 @@
+      self$input <- .createReactiveValues(private$.input, readonly = TRUE)
+    },
+    onFlush = function(fun, once) {
+      if (!isTRUE(once)) {
+        return(private$flushCBs$register(fun))
+      } else {
+        dereg <- private$flushCBs$register(function() {
+          dereg()
+          fun()
+        })
+        return(dereg)
+      }
+    },
+    onFlushed = function(fun, once) {
+      if (!isTRUE(once)) {
+        return(private$flushedCBs$register(fun))
+      } else {
+        dereg <- private$flushedCBs$register(function() {
+          dereg()
+          fun()
+        })
+        return(dereg)
+      }
+    },
+    onEnded = function(sessionEndedCallback) {
+      private$endedCBs$register(sessionEndedCallback)
+    },
+
+    isEnded = function(){ private$closed },
+    isClosed = function(){ private$closed },
+    close = function(){ private$closed <- TRUE },
+
+    #FIXME: this is wrong. Will need to be more complex.
+    cycleStartAction = function(callback){ callback() },
+
+    # Needed for image rendering. Base64-encode the given file.
+    fileUrl = function(name, file, contentType='application/octet-stream') {
+      bytes <- file.info(file)$size
+      if (is.na(bytes))
+        return(NULL)
+
+      fileData <- readBin(file, 'raw', n=bytes)
+      b64 <- rawToBase64(fileData)
+      return(paste('data:', contentType, ';base64,', b64, sep=''))
+    },
+
+    setInputs = function(...) {
+      vals <- list(...)
+      # TODO: is there really not a way to access `names` from inside an lapply?
+      lapply(names(vals), function(k){
+        v <- vals[[k]]
+        private$.input$set(k, v)
+      })
+
+      private$flush()
+    },
+
+    .scheduleTask = function(millis, callback) {
+      id <- private$timer$schedule(millis, callback)
+
+      # Return a deregistration callback
+      function() {
+        invisible(private$timer$unschedule(id))
+      }
+    },
+    elapse = function(millis) {
+      msLeft <- millis
+
+      while (msLeft > 0){
+        t <- private$timer$timeToNextEvent()
+
+        if (is.infinite(t) || t <= 0 || msLeft < t){
+          # Either there's no good upcoming event or we can't make it to it in the allotted time.
+          break
+        }
+        msLeft <- msLeft - t
+        private$timer$elapse(t)
+
+        # timerCallbacks must run before flushReact.
+        private$timer$executeElapsed()
+        private$flush()
+      }
+
+      private$timer$elapse(msLeft)
+
+      # TODO: needed? We're guaranteed to not have anything to run given the above loop, right?
+      private$timer$executeElapsed()
+      private$flush()
+    },
+
+    .now = function() {
+      # Contract is to return Sys.time, which is seconds, not millis.
+      private$timer$getElapsed()/1000
+    },
+
+    defineOutput = function(name, value, label) {
+      obs <- observe({
+        # We could just stash the promise, but we get an "unhandled promise error". This bypasses
+        prom <- NULL
+        tryCatch({
```

Comment:
I think it would be a little clearer if the `tryCatch` only wrapped the `v <- value(self, name)`, and not the `if-else` stuff that checks whether `v` is a promise, since that stuff can't really error.